### DISCUSSION TITLE
Full vs Gateway mode

### DISCUSSION THREAD

**Discussion by faragher** [markqvist/Reticulum#378] [Help & Questions] [answered]
Based on my reading the Full and Gateway modes are functionally identical except Gateway seems to more aggressively find paths for its clients. 

- Is my understanding correct? Path-seeking behavior is the functional difference?
 
- Does it simply repeat the request on other interfaces? 
 
- If so, does that mean a node with only a single interface doesn't have a functional difference between the two modes?
 
- Is there any advantage for, say, a TCPServer to define itself as a gateway and use a full connection to the testnet over the TCPClients?

Trying to wrap my head around the more advanced network infrastructure available.

**Comment by markqvist** [comment, marked as answer]
Your understanding is fundamentally correct:

- An interface is put into gateway mode, when it (the interface itself) acts as a gateway for other peers on the interface.
- The functional difference is that an interface in gateway mode will recursively resolve path requests for other peers on that interface
- If a path request is received on an interface in gateway mode, the following can happen:
  - If a path is already known, the transport instance will immediately answer the path request to the requester.
  - In no path is known, the transport instance will attempt to resolve a path on other interfaces that could potentially have a path to the destination, and wait for a potential reply.
  - If a reply is received before the timeout, the original path request is answered.

This behaviour allows full recursive path discovery over broad and complex network topologies,as long as you remember to set interface modes. For reference, interfaces in Access Point mode also behave this way, in regards to path discovery.

**Reply by faragher** [reply]
Thank you. The similar behavior in the AP seems appropriate.

**Comment by markqvist** [comment]
Forgot this last one:

> Is there any advantage for, say, a TCPServer to define itself as a gateway and use a full connection to the testnet over the TCPClients?

Yes, if for example you intend TCP clients to connect to such an interface to access a broader network (the testnet, for example) this would be a good idea. Since the transport instance hosting the TCP Server interface is permanently connected to the broader network, it will have practically full knowledge of all relevant paths, and any clients connected to it will be able to immediately resolve paths to anything, without having to wait for an announce arrive to themselves first.
